linfa-clustering/src/dbscan/algorithm.rs

```diff
@@ -0,0 +1,65 @@
+use crate::dbscan::hyperparameters::DBScanHyperParams;
+use ndarray::{Array1, ArrayBase, Axis, Data, Ix1, Ix2};
+
+pub fn predict(
+    hyperparameters: DBScanHyperParams,
+    observations: &ArrayBase<impl Data<Elem = f64> + Sync, Ix2>,
+) -> Array1<Option<usize>> {
+    let mut result = Array1::from_elem(observations.dim().1, None);
+    let mut latest_id = 0;
+    for (i, obs) in observations.axis_iter(Axis(1)).enumerate() {
+        if result[i].is_some() {
+            continue;
+        }
+        let n = find_neighbors(&obs, observations, hyperparameters.tolerance());
+        if n.len() < hyperparameters.minimum_points() {
+            continue;
+        }
+        // Now go over the neighbours adding them to the cluster
```

Comment:
Before initialising the `search_queue`, shouldn't we perform the cluster assignment? I.e. `result[i] = latest_id`?
Otherwise `obs` will end up in `search_queue` and we might be duplicating a bunch of work.

Reply:
Basically, as soon as we know that it has enough neighbors.

Reply:
Done :+1:

Reply:
If you put the assignment before `search_queue`, you can avoid `&& **x != i` in the filter statement :grin: